### ISSUE TITLE
SIGSEGV occurred in EVP_MD_CTX_cleanup() immediately after malloc() returned NULL in pkey_hmac_init().

### ISSUE THREAD

**Issue by masanore** [libressl-portable/openbsd#129]
We customize and use libtls under the standard SDK for PlayStation®4 and PlayStation®5.
We appreciate it.

Since malloc fails relatively often on game consoles, our crash reporting system occasionally finds bugs in those situations.
We would like to report one of the bugs that has not been fixed in master yet.

When malloc() fails in pkey_hmac_init(), (EVP_PKEY_CTX*)ctx->data becomes NULL.

https://github.com/libressl-portable/openbsd/blob/2ce413b9f8879a3748e9a8ebf4eefe08d95bd652/src/lib/libcrypto/hmac/hm_pmeth.c#L78-L85

The address of NULL+0x20 is passed as an argument of HMAC_CTX_cleanup() as &hctx->ctx in pkey_hmac_cleanup().
HMAC_CTX_cleanup() passes 0x20 + 0x38 as & ctx->i_ctx as an argument to EVP_MD_CTX_cleanup().
And EVP_MD_CTX_cleanup() causes SIGSEGV when trying to access 0x58 at line 388.
Since the argument is not NULL, HMAC_CTX_cleanup() and EVP_MD_CTX_cleanup() cannot check for NULL, so I think you should check if hctx is not NULL in pkey_hmac_cleanup().
https://github.com/libressl-portable/openbsd/blob/2ce413b9f8879a3748e9a8ebf4eefe08d95bd652/src/lib/libcrypto/hmac/hm_pmeth.c#L120-L125
https://github.com/libressl-portable/openbsd/blob/2ce413b9f8879a3748e9a8ebf4eefe08d95bd652/src/lib/libcrypto/hmac/hmac.c#L235-L238
https://github.com/libressl-portable/openbsd/blob/2ce413b9f8879a3748e9a8ebf4eefe08d95bd652/src/lib/libcrypto/evp/digest.c#L382-L389

They are called from int_ctx_new().
The entity of pmeth->init() is pkey_hmac_init().
Therefore, if malloc() fails at pkey_hmac_init(), it will always be reproduced.

https://github.com/libressl-portable/openbsd/blob/2ce413b9f8879a3748e9a8ebf4eefe08d95bd652/src/lib/libcrypto/evp/pmeth_lib.c#L213-L214

For reference, the backtraces of pkey_hmac_init() and EVP_MD_CTX_cleanup() are as follows.
(When calling with libtls)

### pkey_hmac_init() backtrace
>	tlstest.exe!pkey_hmac_init(evp_pkey_ctx_st * ctx) Line 80	C
 	tlstest.exe!int_ctx_new(evp_pkey_st * pkey, engine_st * e, int id) Line 213	C
 	tlstest.exe!EVP_PKEY_CTX_new_id(int id, engine_st * e) Line 276	C
 	tlstest.exe!EVP_PKEY_new_mac_key(int type, engine_st * e, const unsigned char * key, int keylen) Line 211	C
 	tlstest.exe!tls1_P_hash(const env_md_st * md, const unsigned char * secret, unsigned __int64 secret_len, const void * seed1, unsigned __int64 seed1_len, const void * seed2, unsigned __int64 seed2_len, const void * seed3, unsigned __int64 seed3_len, const void * seed4, unsigned __int64 seed4_len, const void * seed5, unsigned __int64 seed5_len, unsigned char * out, unsigned __int64 out_len) Line 179	C
 	tlstest.exe!tls1_PRF(ssl_st * s, const unsigned char * secret, unsigned __int64 secret_len, const void * seed1, unsigned __int64 seed1_len, const void * seed2, unsigned __int64 seed2_len, const void * seed3, unsigned __int64 seed3_len, const void * seed4, unsigned __int64 seed4_len, const void * seed5, unsigned __int64 seed5_len, unsigned char * out, unsigned __int64 out_len) Line 280	C
 	tlstest.exe!tls12_derive_master_secret(ssl_st * s, unsigned char * premaster_secret, unsigned __int64 premaster_secret_len) Line 108	C
 	tlstest.exe!ssl3_send_client_kex_ecdhe(ssl_st * s, cbb_st * cbb) Line 1913	C
 	tlstest.exe!ssl3_send_client_key_exchange(ssl_st * s) Line 2059	C
 	tlstest.exe!ssl3_connect(ssl_st * s) Line 414	C
 	tlstest.exe!tls13_legacy_connect(ssl_st * ssl) Line 472	C
 	tlstest.exe!SSL_connect(ssl_st * s) Line 992	C
 	tlstest.exe!tls_handshake_client(tls * ctx) Line 451	C
 	tlstest.exe!tls_handshake(tls * ctx) Line 790	C
 	tlstest.exe!tls_write(tls * ctx, const void * buf, unsigned __int64 buflen) Line 848	C
 	tlstest.exe!do_tls_client() Line 585	C

### EVP_MD_CTX_cleanup() backtrace
>	tlstest.exe!EVP_MD_CTX_cleanup(env_md_ctx_st * ctx) Line 388	C
 	tlstest.exe!HMAC_CTX_cleanup(hmac_ctx_st * ctx) Line 239	C
 	tlstest.exe!pkey_hmac_cleanup(evp_pkey_ctx_st * ctx) Line 126	C
 	tlstest.exe!EVP_PKEY_CTX_free(evp_pkey_ctx_st * ctx) Line 343	C
 	tlstest.exe!int_ctx_new(evp_pkey_st * pkey, engine_st * e, int id) Line 215	C
 	tlstest.exe!EVP_PKEY_CTX_new_id(int id, engine_st * e) Line 276	C
 	tlstest.exe!EVP_PKEY_new_mac_key(int type, engine_st * e, const unsigned char * key, int keylen) Line 211	C
 	tlstest.exe!tls1_P_hash(const env_md_st * md, const unsigned char * secret, unsigned __int64 secret_len, const void * seed1, unsigned __int64 seed1_len, const void * seed2, unsigned __int64 seed2_len, const void * seed3, unsigned __int64 seed3_len, const void * seed4, unsigned __int64 seed4_len, const void * seed5, unsigned __int64 seed5_len, unsigned char * out, unsigned __int64 out_len) Line 179	C
 	tlstest.exe!tls1_PRF(ssl_st * s, const unsigned char * secret, unsigned __int64 secret_len, const void * seed1, unsigned __int64 seed1_len, const void * seed2, unsigned __int64 seed2_len, const void * seed3, unsigned __int64 seed3_len, const void * seed4, unsigned __int64 seed4_len, const void * seed5, unsigned __int64 seed5_len, unsigned char * out, unsigned __int64 out_len) Line 280	C
 	tlstest.exe!tls12_derive_master_secret(ssl_st * s, unsigned char * premaster_secret, unsigned __int64 premaster_secret_len) Line 108	C
 	tlstest.exe!ssl3_send_client_kex_ecdhe(ssl_st * s, cbb_st * cbb) Line 1913	C
 	tlstest.exe!ssl3_send_client_key_exchange(ssl_st * s) Line 2059	C
 	tlstest.exe!ssl3_connect(ssl_st * s) Line 414	C
 	tlstest.exe!tls13_legacy_connect(ssl_st * ssl) Line 472	C
 	tlstest.exe!SSL_connect(ssl_st * s) Line 992	C
 	tlstest.exe!tls_handshake_client(tls * ctx) Line 451	C
 	tlstest.exe!tls_handshake(tls * ctx) Line 790	C
 	tlstest.exe!tls_write(tls * ctx, const void * buf, unsigned __int64 buflen) Line 848	C

### Reproduction method
Since it is difficult to make malloc fail only in the relevant part, it can be reproduced by compiling the rewritten libressl as shown below and executing the following command.
`./apps/openssl/openssl s_client -connect example.com:443 -no_tls1_3`

Note that this does not occur in TLS 1.3 because pkey_hmac_init() is not used.

```
diff --git a/src/lib/libcrypto/hmac/hm_pmeth.c b/src/lib/libcrypto/hmac/hm_pmeth.c
index 8964c6487..51936e79e 100644
--- a/src/lib/libcrypto/hmac/hm_pmeth.c
+++ b/src/lib/libcrypto/hmac/hm_pmeth.c
@@ -80,7 +80,9 @@ pkey_hmac_init(EVP_PKEY_CTX *ctx)
 {
        HMAC_PKEY_CTX *hctx;

-       hctx = malloc(sizeof(HMAC_PKEY_CTX));
+       // for test
+       //hctx = malloc(sizeof(HMAC_PKEY_CTX));
+       hctx = NULL;
        if (!hctx)
                return 0;
        hctx->md = NULL;

```
### LibreSSL version
This issue is reproducible on LibreSSL master.

### OS
 RockyLinux8, Windows10

**Comment by kinichiro** [comment]
@masanore Thanks for your report.
Upstream was fixed by https://github.com/openbsd/src/commit/f802266ee2c993ad1c979c057f92717755cafb58 .
Could you see if this solves this issue, please ?

**Comment by masanore** [comment]
@kinichiro Thanks for the the quick fix.
I actually tried to execute it after correction, but there was no problem.
And I didn't notice about gost, but it certainly had a similar problem.
I made the calloc in pkey_gost01_init () fail in the same way and executed the following command to confirm the correction. It was fixed correctly.
`./apps/openssl/openssl s_client -connect tlsgost-2001.cryptopro.ru:443 -servername tlsgost-2001.cryptopro.ru -cipher GOST2012256-GOST89-GOST89 -no_tls1_3`
Thank you.
